### PR TITLE
Add `raft::stats::neighborhood_recall`

### PR DESCRIPTION
**Pull request by divyegala**
This PR adds a primitive to compute recall score for Nearest Neighbor Algorithms

### CODE REVIEW

File: cpp/include/raft/stats/neighborhood_recall.cuh [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+  DistanceValueType eps_val = 0.001;
+  if (eps.has_value()) { eps_val = *eps.value().data_handle(); }
+
+  detail::neighborhood_recall(
+    res, indices, ref_indices, distances, ref_distances, recall_score, eps_val);
+}
+
+/**
+ * @brief Calculate Neighborhood Recall score on the host for indices, distances computed by any
```
Comment (cjnolet) [inline cpp/include/raft/stats/neighborhood_recall.cuh:101]: New functions need usage examples.

File: cpp/include/raft/stats/neighborhood_recall.cuh [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+ * @{
+ */
+
+/**
+ * @brief Calculate Neighborhood Recall score on the device for indices, distances computed by any
+ * Nearest Neighbors Algorithm against reference indices, distances. Recall score is calculated by
+ * comparing the total number of matching indices and dividing that value by the total size of the
+ * indices matrix of dimensions (D, k). If distance matrices are provided, then non-matching indices
+ * could be considered a match if abs(dist, ref_dist) < eps.
+ *
```
Comment (cjnolet) [inline cpp/include/raft/stats/neighborhood_recall.cuh:45]: Usage example

**Review by cjnolet — CHANGES_REQUESTED** [review]

File: cpp/include/raft/stats/detail/neighborhood_recall.cuh [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  for (IndexType col_idx = lane_idx; col_idx < indices.extent(1); col_idx += 32) {
+    for (IndexType ref_col_idx = 0; ref_col_idx < ref_indices.extent(1); ref_col_idx++) {
+      if (indices(row_idx, col_idx) == ref_indices(row_idx, ref_col_idx)) {
+        thread_recall_score += 1;
+        break;
+      } else if (distances.has_value()) {
+        auto dist               = distances.value()(row_idx, col_idx);
+        auto ref_dist           = ref_distances.value()(row_idx, ref_col_idx);
+        DistanceValueType diff  = raft::abs(dist - ref_dist);
+        DistanceValueType m     = std::max(std::abs(dist), std::abs(ref_dist));
```
[the remark this answers is not in the document]
Comment (divyegala) [reply cpp/include/raft/stats/detail/neighborhood_recall.cuh:64]: Good point!

**Review by lowener — APPROVED** [review]
LGTM

**Review by cjnolet — APPROVED** [review]

**Pull request merged**